src/ModelContextProtocol/Authentication/AuthorizationHelpers.cs

```diff
@@ -0,0 +1,165 @@
+    public static async Task<ProtectedResourceMetadata> ExtractProtectedResourceMetadata(
+        HttpResponseMessage response,
+        Uri serverUrl,
+        CancellationToken cancellationToken = default)
+    {
+        if (response.StatusCode != System.Net.HttpStatusCode.Unauthorized)
+        {
+            throw new InvalidOperationException($"Expected a 401 Unauthorized response, but received {(int)response.StatusCode} {response.StatusCode}");
+        }
+
+        // Extract the WWW-Authenticate header
+        if (!response.Headers.WwwAuthenticate.Any())
+        {
+            throw new InvalidOperationException("The 401 response does not contain a WWW-Authenticate header");
+        }
+
+        // Look for the Bearer authentication scheme with resource_metadata parameter
+        string? resourceMetadataUrl = null;
+        foreach (var header in response.Headers.WwwAuthenticate)
+        {
+            if (header.Scheme.Equals("Bearer", StringComparison.OrdinalIgnoreCase))
+            {
+                var parameters = header.Parameter;
+                if (string.IsNullOrEmpty(parameters))
+                {
+                    continue;
+                }
+
+                // Parse the parameters to find resource_metadata
+                resourceMetadataUrl = ParseWwwAuthenticateParameters(parameters, "resource_metadata");
+                if (resourceMetadataUrl != null)
+                {
+                    break;
+                }
+            }
+        }
+
+        if (resourceMetadataUrl == null)
+        {
+            throw new InvalidOperationException("The WWW-Authenticate header does not contain a resource_metadata parameter");
+        }
+
+        Uri metadataUri = new(resourceMetadataUrl);
+        
+        var metadata = await FetchProtectedResourceMetadataAsync(metadataUri, cancellationToken) ?? throw new InvalidOperationException($"Failed to fetch resource metadata from {resourceMetadataUrl}");
+        if (!VerifyResourceMatch(metadata, serverUrl))
+        {
+            throw new InvalidOperationException(
+                $"Resource URI in metadata ({metadata.Resource}) does not match the server URI ({serverUrl})");
+        }
+
+        return metadata;
+    }
+
+    /// <summary>
+    /// Parses the WWW-Authenticate header parameters to extract a specific parameter.
+    /// </summary>
+    /// <param name="parameters">The parameter string from the WWW-Authenticate header.</param>
+    /// <param name="parameterName">The name of the parameter to extract.</param>
+    /// <returns>The value of the parameter, or null if not found.</returns>
+    private static string? ParseWwwAuthenticateParameters(string parameters, string parameterName)
+    {
+        // Handle parameters in the format: param1="value1", param2="value2"
+        var paramDict = parameters.Split(',')
```

Comment:
This linq code is extremely inefficient. Please reimplement without 4 reallocations of the entire structure. It can be done with a single dictionary allocation after the split call.